osu.Game/Rulesets/Judgements/JudgementResult.cs

```diff
@@ -33,16 +33,19 @@ public class JudgementResult
         public readonly Judgement Judgement;
 
         /// <summary>
-        /// The offset from a perfect hit at which this <see cref="JudgementResult"/> occurred.
+        /// The offset of <see cref="TimeAbsolute"/> from the end time of <see cref="HitObject"/>, clamped by <see cref="DrawableHitObject.MaximumJudgementOffset"/>.
         /// Populated when this <see cref="JudgementResult"/> is applied via <see cref="DrawableHitObject.ApplyResult"/>.
         /// </summary>
         public double TimeOffset { get; internal set; }
 
         /// <summary>
         /// The absolute time at which this <see cref="JudgementResult"/> occurred.
-        /// Equal to the (end) time of the <see cref="HitObject"/> + <see cref="TimeOffset"/>.
+        /// Populated when this <see cref="JudgementResult"/> is applied via <see cref="DrawableHitObject.ApplyResult"/>.
         /// </summary>
-        public double TimeAbsolute => HitObject.GetEndTime() + TimeOffset;
+        /// <remarks>
+        /// This is initially set to the end time of <see cref="HitObject"/>.
+        /// </remarks>
+        public double TimeAbsolute { get; internal set; }
```

Comment:
Am I the only one that's completely spooked that `TimeOffset` and `TimeAbsolute` are both independently settable with this, and therefore can completely desync from one another?

I've written [a branch](https://github.com/ppy/osu/compare/master...bdach:dho-store-absolute-time-2?expand=1) that would have both settable before, but with the caveat that setting one would also propagate to the other. Inverting the relationship and having `TimeOffset` be derived from `TimeAbsolute` [was also floated at one point](https://github.com/ppy/osu/issues/17482#issuecomment-1084351300), but that [is more work](https://github.com/ppy/osu/issues/17482#issuecomment-1084365647).

Reply:
I also concerned about that, but concluded this is okay for now because it is `internal set`.
Ideally we don't want to use such `internal` properties for such a distant coupling IMO but it seems difficult to change it without much refactoring.

Reply:
@peppy just to confirm, did you see this when reviewing prior to approving? Like to me this is very scary, even if it's internal. We can also break this too.

Would like @smoogipoo's opinion on this too while we're at it.

Reply:
Yep. I'm quite fine with this. I don't mind cases like this because it's hard coupled at one or two points of interaction. I have faith that if that ever changed one of us would raise alarm bells and we'd reconsider at that point.

Reply:
Okay then. In that case, I self requested re-review as I will want to look through all cases where either property is set to make sure that the other doesn't contain a value that's invalid for reading.